### PR TITLE
Add ColorBar.display_low and display_high

### LINKED ISSUE #12183: [BUG] EqHistColorMapper with rescale_discrete_levels=True should not show colorbar labels outside the data limits
### Software versions

Python version      :  3.9.7 | packaged by conda-forge | (default, Sep 29 2021, 19:20:46) 
IPython version     :  7.30.1
Tornado version     :  6.1
Bokeh version       :  3.0.0dev7+4.g4a2199cd2
BokehJS static path :  /home/iant/github/bokeh/bokeh/server/static
node.js version     :  v16.12.0
npm version         :  8.1.0
Operating system    :  Linux-5.11.0-37-generic-x86_64-with-glibc2.31

### Browser name and version

Not browser-specific

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

If using `EqHistColorMapper(..., rescale_discrete_levels=True)` and there are fewer than 100 discrete levels in the data values then the lower limit is decreased so that the values are rendered towards the top end of the palette. When this occurs, the colors and levels that are not used are still drawn in the `ColorBar` and this is confusing as the values are meaningless. For example, when used by datashader using a `count` per pixel values will always be zero or positive integers, but the labels of the unused colors show negative values.

### Observed behavior

See screenshot below.

### Example code

```Python
from bokeh.io import show
from bokeh.models import ColorBar, EqHistColorMapper
from bokeh.palettes import Blues9
from bokeh.plotting import figure
import numpy as np

n = 100
im = np.zeros((n, n), dtype=np.uint32)
for j in range(2, 4):
    im[:n//j, :n//j] = j-1

p = figure(width=400, height=300)
color_mapper = EqHistColorMapper(Blues9[::-1], rescale_discrete_levels=True)

p.image(image=[im], x=0, y=0, dw=10, dh=10, color_mapper=color_mapper)

color_bar = ColorBar(color_mapper=color_mapper)
p.add_layout(color_bar, 'right')

show(p)
```

### Stack traceback or browser console output

_No response_

### Screenshots

before

### PR DESCRIPTION
**Pull request by ianthomas23**
- [x] issues: partially fixes #12183
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

This PR adds `low_cutoff` and `high_cutoff` to `ColorBar`. It supersedes PR #12184 and adds the new functionality required to fix issue #12183 but the actual fix will follow in a separate PR that will programmatically set the `low_cutoff` as required.

Note that only the color bar output is cutoff, data that is colored in a plot using a `ColorMapper` that is the subject of `ColorBar` cutoffs is unaffected and will use the whole color range (unless requested not to in the usual way).

Demo:

cutoff

using code
```python
from bokeh.io import show
from bokeh.layouts import column
from bokeh.models import ColorBar, ColumnDataSource, Slider, LinearColorMapper
from bokeh.palettes import Spectral10
from bokeh.plotting import figure
import numpy as np

p = figure(width=500, height=300)
color_mapper = LinearColorMapper(Spectral10)

x = np.arange(1, 11)
cds = ColumnDataSource(data=dict(x=x, y=x, v=x))
c = p.circle(source=cds, x="x", y="y", size=40, fill_color=dict(field="v", transform=color_mapper))

color_bar = ColorBar(color_mapper=color_mapper, bar_line_color="black")
p.add_layout(color_bar, "right")

low_slider = Slider(start=-1, end=10, step=0.1, value=-1, title="low cutoff")
low_slider.js_link("value", color_bar, "low_cutoff")

high_slider = Slider(start=-1, end=10, step=0.1, value=10, title="high cutoff")
high_slider.js_link("value", color_bar, "high_cutoff")

show(column(p, low_slider, high_slider))
```
The whole of the colored rectangle corresponding to the cutoff value is displayed, so there won't just be a single row of colored pixels for example. If `high_cutoff < low_cutoff` or `high_cutoff` is lower than the minimum data value or `low_cutoff` is higher than the maximum data value then no colors are displayed.

Implementation details:

1. There is separate handling of EqHist and Linear/Log color mapper classes in `ColorBar`, as was the case before this PR.
2. `ColorBar._image` can now be `null` if there is no image to display.
3. There are extra tweaks in some `ColorMapper`, `Ticker` and `Axis` classes to deal with empty data.
4. The `ContinuousColorMapper`-derived classes now have separate functions for transforms from data value to palette indices and back again. Some of the code for this was originally in the `cmap` functions. The `cmap` function is now solely in `ContinuousColorMapper` as it doesn't need to be overridden. It also no longer takes the `scan_data` as an argument as it was always recalculated and stored as `this._scan_data` before `cmap` was called, so I am just reading it from `this._scan_data`.

(Edited to tick release notes check box)

### CODE REVIEW

**Comment by ianthomas23** [comment]
I need to rebase this and deal with the conflicts in the colorbar code caused by the contouring PR.

**Comment by ianthomas23** [comment]
I have rebased this.

**Comment by bryevdv** [comment]
The failure is for bundle size 
```
✗ js/bokeh.min.js          851 kB >   850 kB
```

**Comment by ianthomas23** [comment]
> The failure is for bundle size

I've increase the bundle size from 850 to 900 kB as it looks like we have historically used 50 kB increments.

File: bokeh/models/annotations/legends.py [hunk 1]
```diff
@@ -234,6 +234,16 @@ def __init__(self, *args, **kwargs) -> None:
...
 
+    low_cutoff = Nullable(Float, help="""
+    The lowest value to show in the color bar. The whole of the color entry
+    containing this value is shown.
+    """)
+
+    high_cutoff = Nullable(Float, help="""
+    The highest value to show in the color bar. The whole of the color entry
+    containing this value is shown.
+    """)
```
Comment (bryevdv) [inline bokeh/models/annotations/legends.py:245]: Would it make sense to use names like `display_low` and `display_high` (or similar) to help reinforce that these properties only affect the visual presentation, and not the color mapping?

Comment (ianthomas23) [reply bokeh/models/annotations/legends.py:245]: Yes, those are better names. I've also changed `low_index` and `high_index` to `index_low` and `index_high` to better associate them with `display_low` and `display_high`.

**Comment by ianthomas23** [comment]
I have added a release note as this an API addition. I've added it to the `API changes` section although maybe it should go under `Additions` instead?

**Comment by mattpap** [comment]
I would add a test or two for the ability to update bounds.

**Comment by ianthomas23** [comment]
> I would add a test or two for the ability to update bounds.

I am glad you suggested this, it revealed a bug where I wasn't handling vertical and horizontal colorbars the same.

I've rebased it to pick up the baseline image changes, added two update tests for horizontal/vertical EqHist/Linear color mappers, and fixed the bug.

**Pull request merged**